### PR TITLE
Implement Zstandard streaming decompression

### LINKED ISSUE #46: The returned value from ZSTD_getFrameContentSize is not checked.
The return value of `ZSTD_getFrameContentSize` is not checked before being passed to `malloc`.
https://github.com/manzt/numcodecs.js/blob/b9a8ca932240a6de9ad6c00f9820ab5e6bbca4c1/codecs/zstd/zstd_codec.cpp#L28-L29

Here is the relevant excerpt from the [Zstd manual](https://facebook.github.io/zstd/zstd_manual.html) as of version 1.5.1

> #define ZSTD_CONTENTSIZE_UNKNOWN (0ULL - 1)
> #define ZSTD_CONTENTSIZE_ERROR   (0ULL - 2)
> unsigned long long ZSTD_getFrameContentSize(const void *src, size_t srcSize);
>   `src` should point to the start of a ZSTD encoded frame.
>   `srcSize` must be at least as large as the frame header.
>             hint : any size >= `ZSTD_frameHeaderSize_max` is large enough.
>   @return : - decompressed size of `src` frame content, if known
>             - ZSTD_CONTENTSIZE_UNKNOWN if the size cannot be determined
>             - ZSTD_CONTENTSIZE_ERROR if an error occurred (e.g. invalid magic number, srcSize too small)
>    note 1 : a 0 return value means the frame is valid but "empty".
>    note 2 : decompressed size is an optional field, it may not be present, typically in streaming mode.
>             When `return==ZSTD_CONTENTSIZE_UNKNOWN`, data to decompress could be any size.
>             In which case, it's necessary to use streaming mode to decompress data.
>             Optionally, application can rely on some implicit limit,
>             as ZSTD_decompress() only needs an upper bound of decompressed size.
>             (For example, data could be necessarily cut into blocks <= 16 KB).
>    note 3 : decompressed size is always present when compression is completed using single-pass functions,
>             such as ZSTD_compress(), ZSTD_compressCCtx() ZSTD_compress_usingDict() or ZSTD_compress_usingCDict().
>    note 4 : decompressed size can be very large (64-bits value),
>             potentially larger than what local system can handle as a single memory segment.
>             In which case, it's necessary to use streaming mode to decompress data.
>    note 5 : If source is untrusted, decompressed size could be wrong or intentionally modified.
>             Always ensure return value fits within application's authorized limits.
>             Each application can set its own limits.
>    note 6 : This function replaces ZSTD_getDecompressedSize() 
> 

The return value from `ZSTD_getFrameContentSize` could be either `ZSTD_CONTENTSIZE_UNKNOWN` or `ZSTD_CONTENTSIZE_ERROR`. Preferably the value should be bounded by the _expected_ destination buffer size rather than letting this be unconstrained from the data.

xref: https://github.com/google/neuroglancer/issues/625

### PR DESCRIPTION
**Pull request by mkitti**
The output of ZSTD_getFrameContentSize is now checked.

If ZSTD_CONTENTSIZE_UNKNOWN, then we try to stream decompress into a 1 MiB output buffer
that grows in <strike>1 MiB</strike> 128 KiB increments

We use `WebAssembly.Exception` to report errors.

Fix #46

### CODE REVIEW

**Comment by manzt** [comment]
Is this ready for review?

File: codecs/zstd/build.sh [hunk 1]
```diff
@@ -49,7 +49,7 @@ cd ../../../
     ${OPTIMIZE} \
     -I "$CODEC_DIR/lib" \
     --closure 1 \
-    -fwasm-exceptions
+    -fwasm-exceptions \
```
Comment (manzt) [inline codecs/zstd/build.sh:52]: Ah, I see you did use wasm exceptions!

Comment (manzt) [reply codecs/zstd/build.sh:52]: I believe this is listed as "legacy" now, but the alternative proposal isn't supported in any runtimme at all so I'm good with using for now: https://webassembly.org/features/

**Comment by mkitti** [comment]
> Is this ready for review?

Almost. I got a few notes from Janelia Scientific Computing Code Review this morning, so I have a few changes to make this evening.

I also think a little more work is needed to catch the exceptions properly.

I'm slightly unclear which exception handling feature I'm using on that chart.

**Comment by mkitti** [comment]
According to Mozilla, all major browsers have supported this since 2021 or 2022.

image

File: src/zstd.ts [hunk 2]
```diff
@@ -52,14 +52,18 @@ const Zstd: CodecConstructor<ZstdConfig> = class Zstd implements Codec {
...
+      const view = module.decompress(data);
+      const result = new Uint8Array(view); // Copy view and free wasm memory
+      module.free_result();
+      if (out !== undefined) {
+        out.set(result);
+        return out;
+      }
+      return result;
+    } catch (err) {
+      throw new Error(module.getExceptionMessage(err).toString());
```
Comment (manzt) [inline src/zstd.ts:65]: Nice! Thanks for this.

File: codecs/zstd/zstd_codec.cpp [hunk 3]
```diff
@@ -20,16 +20,114 @@ val compress(std::string source, int level)
...
-    dest_ptr = (char *)malloc((size_t)dest_size);
+    unsigned long long dest_size = ZSTD_getFrameContentSize(source_ptr, source_size);
+
+    // If Zstd_compressStream was used, we may not know the frame content size.
+    // https://github.com/manzt/numcodecs.js/issues/46
+    if (dest_size == ZSTD_CONTENTSIZE_UNKNOWN) {
+        // guess decompressed buffer size based on source size
+        dest_size = source_size*2;
+
+        // Initialize the destination size to 1 MiB at minimum
```
Comment (manzt) [inline codecs/zstd/zstd_codec.cpp:60]:
```suggestion
        // Initialize the destination size to 128 KiB at minimum
```

**Review by manzt — APPROVED** [review]
Awesome, just some minor comments! Thanks for looking in to the exceptions and the nice tests!

File: codecs/zstd/zstd_codec.cpp (hunk 3 above)
Comment (mkitti) [inline codecs/zstd/zstd_codec.cpp:60]:
```suggestion
        // Initialize the destination size to DEST_GROWTH_SIZE (default: 128 KiB) at minimum
```

**Comment by manzt** [comment]
Thanks! Released in v0.3.2 https://github.com/manzt/numcodecs.js/releases/tag/v0.3.2

**Pull request merged**